src/stream.rs

```diff
@@ -0,0 +1,293 @@
+                        name: unit_state.unit.name.clone(),
+                    })))
+                    .await?;
+
+                unit_state.is_gone = true;
+                continue;
+            }
+            Err(err) => return Err(err.into()),
+        }
+    }
+
+    // remove state for all units that are gone
+    state.units.retain(|_, v| !v.is_gone);
+
+    Ok(())
+}
+
+/// The last know information about a unit and various other information to track whether it is
+/// worth checking the unit for updates or not.
+struct UnitState {
+    unit: Unit,
+    backoff: Duration,
+    last_checked: Instant,
+    last_changed: Instant,
+    is_gone: bool,
+}
+
+impl UnitState {
+    fn new(unit: Unit) -> Self {
+        Self {
+            unit,
+            backoff: Duration::ZERO,
+            last_checked: Instant::now(),
+            last_changed: Instant::now(),
+            is_gone: false,
+        }
+    }
+
+    /// Whether the unit should be checked for updates or not. This can be used to check stationary
+    /// units less often.
+    fn should_update(&self) -> bool {
+        self.last_checked.elapsed() >= self.backoff
+    }
+
+    /// Check the unit for updates and return whether the unit got changed or not.
+    async fn update(&mut self, ctx: &Context) -> Result<bool, Status> {
+        let mut changed = false;
+
+        // update position
+        let position = Units::get_position(
+            &ctx.rpc,
+            Request::new(UnitName {
+                name: self.unit.name.clone(),
+            }),
+        )
+        .await?
+        .into_inner()
+        .position;
+        if let (Some(a), Some(b)) = (position.as_ref(), self.unit.position.as_ref()) {
+            if !eq(a, b) {
+                changed = true;
+                self.unit.position = position;
+            }
+        }
+
+        // update player name
+        let player_name = Units::get_player_name(
+            &ctx.rpc,
+            Request::new(UnitName {
+                name: self.unit.name.clone(),
+            }),
+        )
+        .await?
+        .into_inner()
+        .player_name;
+        if player_name != self.unit.player_name {
+            changed = true;
+            self.unit.player_name = player_name;
+        }
+
+        // keep track of when it was last checked and changed and determine a corresponding backoff
+        self.last_checked = Instant::now();
+        if changed {
+            self.last_changed = Instant::now();
+            self.backoff = Duration::ZERO;
+        } else {
+            self.backoff = if self.backoff.is_zero() {
+                ctx.poll_rate
+            } else {
+                (self.backoff * 2).min(ctx.max_backoff)
+            }
+        }
+
+        Ok(changed)
+    }
+}
+
+/// Check whether two positions are equal, taking an epsilon into account.
+fn eq(a: &Position, b: &Position) -> bool {
+    const LL_EPSILON: f64 = 0.000001;
+    const ALT_EPSILON: f64 = 0.001;
```

Comment:
Wdyt about the `Position` comparison here. Should we myb increase or decrease the precision of the comparison? I think for the altitude an epsilon of `0.01` might be perfectly fine too.

Reply:
I will do some testing this weekend and see if there is anything to suggest that we need to tune these values.